File: backend/src/modules/auth/services/GoogleOAuthService.ts
import { Request } from 'express'
import { sign, verify } from 'jsonwebtoken'
import { prisma } from '../../../config/database'
import { env } from '../../../config/env'
import { AppError } from '../../../shared/errors/AppError'

const GOOGLE_OAUTH_SCOPES = [
  'https://www.googleapis.com/auth/gmail.readonly',
  'https://www.googleapis.com/auth/gmail.send',
  'https://www.googleapis.com/auth/gmail.modify',
].join(' ')

type GoogleOAuthStatePayload = {
  type: 'gmail_oauth'
  clinicId: string
  userId: string
  returnUrl?: string
  allowedOrigin?: string
}

type GoogleOAuthCredentials = {
  clientId: string
  clientSecret: string
}

function tryExtractCredentialValue(rawValue: string, key: 'clientId' | 'clientSecret') {
  const trimmed = rawValue.trim()
  if (!trimmed) {
    return ''
  }

  const regex = key === 'clientId'
    ? /client[_-]?id["']?\s*[:=]\s*["']([^"']+)["']/i
    : /client[_-]?secret["']?\s*[:=]\s*["']([^"']+)["']/i

  const directMatch = trimmed.match(regex)
  if (directMatch?.[1]) {
    return directMatch[1].trim()
  }

  try {
    const parsed = JSON.parse(trimmed) as any
    const source = parsed?.web ?? parsed?.installed ?? parsed

    if (key === 'clientId') {
      return String(source?.client_id ?? source?.clientId ?? '').trim()
    }

    return String(source?.client_secret ?? source?.clientSecret ?? '').trim()
  } catch {
    return ''
  }
}

function normalizeCredentialValue(rawValue: string | null | undefined, key: 'clientId' | 'clientSecret') {
  const input = String(rawValue ?? '').trim()
  if (!input) {
    return ''
  }

  const extracted = tryExtractCredentialValue(input, key)
  return (extracted || input)
    .trim()
    .replace(/^[\s"'`]+|[\s"',`]+$/g, '')
    .trim()
}

function getNormalizedGoogleCredentials(settings: { gmailClientId?: string | null; gmailClientSecret?: string | null }) {
  const clientId = normalizeCredentialValue(settings.gmailClientId, 'clientId')
  const clientSecret = normalizeCredentialValue(settings.gmailClientSecret, 'clientSecret')

  if (!clientId || !clientSecret) {
    throw new AppError('Salve Client ID e Client Secret do Gmail antes de autenticar', 422)
  }

  if (!clientId.includes('.apps.googleusercontent.com')) {
    throw new AppError(
      'Client ID do Google parece invalido. Use um OAuth Client ID do tipo "Aplicativo Web" e cole apenas o valor do client_id.',
      422,
    )
  }

  return { clientId, clientSecret } satisfies GoogleOAuthCredentials
}

function normalizePublicApiUrl(url: string) {
  return url
    .trim()
    .replace(/\/$/, '')
    .replace(/\/api$/, '')
}

function resolveRequestBaseUrl(req: Request) {
  if (process.env.PUBLIC_API_URL) {
    return normalizePublicApiUrl(process.env.PUBLIC_API_URL)
  }

  const protocol = (req.get('x-forwarded-proto') || req.protocol || 'http').split(',')[0].trim()
  const host = (req.get('x-forwarded-host') || req.get('host') || '').split(',')[0].trim()

  if (!host) {
    throw new AppError('Nao foi possivel determinar a URL publica da API', 500)
  }

  return `${protocol}://${host}`
}

function resolveAllowedOrigin(req: Request, returnUrl?: string) {
  const originHeader = req.get('origin')
  if (originHeader) {
    return originHeader
  }

  const refererHeader = req.get('referer')
  if (refererHeader) {
    try {
      return new URL(refererHeader).origin
    } catch {
    }
  }

  if (returnUrl) {
    try {
      return new URL(returnUrl).origin
    } catch {
    }
  }

  return undefined
}

function buildSafeReturnUrl(returnUrl: string | undefined, allowedOrigin?: string) {
  const fallbackPath = '/admin/configuracoes?tab=integrations'

  if (!returnUrl) {
    return allowedOrigin ? `${allowedOrigin}${fallbackPath}` : fallbackPath
  }

  if (returnUrl.startsWith('/')) {
    return allowedOrigin ? `${allowedOrigin}${returnUrl}` : returnUrl
  }

  try {
    const parsed = new URL(returnUrl)
    if (allowedOrigin && parsed.origin !== allowedOrigin) {
      return `${allowedOrigin}${fallbackPath}`
    }

    return parsed.toString()
  } catch {
    return allowedOrigin ? `${allowedOrigin}${fallbackPath}` : fallbackPath
  }
}

function appendRedirectParams(targetUrl: string, params: Record<string, string>) {
  const isAbsolute = /^https?:\/\//i.test(targetUrl)
  const base = isAbsolute ? undefined : 'http://local'
  const url = new URL(targetUrl, base)

  for (const [key, value] of Object.entries(params)) {
    url.searchParams.set(key, value)
  }

  if (isAbsolute) {
    return url.toString()
  }

  return `${url.pathname}${url.search}${url.hash}`
}

async function exchangeCodeForTokens(input: {
  code: string
  clientId: string
  clientSecret: string
  redirectUri: string
}) {
  const response = await fetch('https://oauth2.googleapis.com/token', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: new URLSearchParams({
      code: input.code,
      client_id: input.clientId,
      client_secret: input.clientSecret,
      redirect_uri: input.redirectUri,
      grant_type: 'authorization_code',
    }),
  })

  const data = await response.json() as {
    access_token?: string
    refresh_token?: string
    expires_in?: number
    error?: string
    error_description?: string
  }

  if (!response.ok || !data.access_token) {
    throw new AppError(data.error_description || data.error || 'Falha ao trocar o codigo OAuth do Google', 400)
  }

  return data
}

async function refreshAccessToken(input: {
  clientId: string
  clientSecret: string
  refreshToken: string
}) {
  const response = await fetch('https://oauth2.googleapis.com/token', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: new URLSearchParams({
      client_id: input.clientId,
      client_secret: input.clientSecret,
      refresh_token: input.refreshToken,
      grant_type: 'refresh_token',
    }),
  })

  const data = await response.json() as {
    access_token?: string
    expires_in?: number
    error?: string
    error_description?: string
  }

  if (!response.ok || !data.access_token) {
    throw new AppError(data.error_description || data.error || 'Falha ao renovar o token do Gmail', 400)
  }

  return data
}

export class GoogleOAuthService {
  async createAuthorizationUrl(req: Request, input: { clinicId: string; userId: string; returnUrl?: string }) {
    if (!input.clinicId) {
      throw new AppError('clinicId e obrigatorio para iniciar a autenticacao do Gmail', 400)
    }

    const membership = await prisma.systemUser.findUnique({
      where: {
        clinicId_userId: {
          clinicId: input.clinicId,
          userId: input.userId,
        },
      },
    })

    if (!membership || !membership.active) {
      throw new AppError('Voce nao tem acesso a esta clinica', 403)
    }

    const settings = await prisma.integrationSettings.findUnique({
      where: { clinicId: input.clinicId },
    })

    const credentials = getNormalizedGoogleCredentials(settings ?? {})
    const redirectUri = `${resolveRequestBaseUrl(req)}/api/auth/google/callback`
    const allowedOrigin = resolveAllowedOrigin(req, input.returnUrl)
    const returnUrl = buildSafeReturnUrl(input.returnUrl, allowedOrigin)

    const state = sign(
      {
        type: 'gmail_oauth',
        clinicId: input.clinicId,
        userId: input.userId,
        returnUrl,
        allowedOrigin,
      } satisfies GoogleOAuthStatePayload,
      env.JWT_SECRET,
      { expiresIn: '15m' },
    )

    const authUrl = `https://accounts.google.com/o/oauth2/v2/auth?${new URLSearchParams({
      client_id: credentials.clientId,
      redirect_uri: redirectUri,
      response_type: 'code',
      access_type: 'offline',
      prompt: 'consent',
      include_granted_scopes: 'true',
      scope: GOOGLE_OAUTH_SCOPES,
      state,
    }).toString()}`

    return { authUrl, redirectUri }
  }

  async handleCallback(req: Request) {
    const stateToken = typeof req.query.state === 'string' ? req.query.state : undefined
    if (!stateToken) {
      throw new AppError('State OAuth ausente', 400)
    }

    const state = verify(stateToken, env.JWT_SECRET) as GoogleOAuthStatePayload
    if (state.type !== 'gmail_oauth') {
      throw new AppError('State OAuth invalido', 400)
    }

    const safeReturnUrl = buildSafeReturnUrl(state.returnUrl, state.allowedOrigin)
    const googleError = typeof req.query.error === 'string' ? req.query.error : undefined
    const googleErrorDescription = typeof req.query.error_description === 'string' ? req.query.error_description : undefined

    if (googleError) {
      return {
        redirectUrl: appendRedirectParams(safeReturnUrl, {
          tab: 'integrations',
          gmail_oauth: 'error',
          gmail_message: googleErrorDescription || googleError,
        }),
      }
    }

    const code = typeof req.query.code === 'string' ? req.query.code : undefined
    if (!code) {
      return {
        redirectUrl: appendRedirectParams(safeReturnUrl, {
          tab: 'integrations',
          gmail_oauth: 'error',
          gmail_message: 'Codigo OAuth do Google ausente',
        }),
      }
    }

    try {
      const membership = await prisma.systemUser.findUnique({
        where: {
          clinicId_userId: {
            clinicId: state.clinicId,
            userId: state.userId,
          },
        },
      })

      if (!membership || !membership.active) {
        throw new AppError('Voce nao tem mais acesso a esta clinica', 403)
      }

      const settings = await prisma.integrationSettings.findUnique({
        where: { clinicId: state.clinicId },
      })

      const credentials = getNormalizedGoogleCredentials(settings ?? {})
      const redirectUri = `${resolveRequestBaseUrl(req)}/api/auth/google/callback`
      const tokens = await exchangeCodeForTokens({
        code,
        clientId: credentials.clientId,
        clientSecret: credentials.clientSecret,
        redirectUri,
      })

      await prisma.integrationSettings.upsert({
        where: { clinicId: state.clinicId },
        update: {
          gmailClientId: credentials.clientId,
          gmailClientSecret: credentials.clientSecret,
          gmailAccessToken: tokens.access_token,
          gmailRefreshToken: tokens.refresh_token || settings?.gmailRefreshToken,
          gmailTokenExpiresAt: tokens.expires_in ? new Date(Date.now() + tokens.expires_in * 1000) : null,
          gmailConnected: true,
        },
        create: {
          clinicId: state.clinicId,
          gmailClientId: credentials.clientId,
          gmailClientSecret: credentials.clientSecret,
          gmailAccessToken: tokens.access_token,
          gmailRefreshToken: tokens.refresh_token,
          gmailTokenExpiresAt: tokens.expires_in ? new Date(Date.now() + tokens.expires_in * 1000) : null,
          gmailConnected: true,
        },
      })

      return {
        redirectUrl: appendRedirectParams(safeReturnUrl, {
          tab: 'integrations',
          gmail_oauth: 'success',
          gmail_message: 'Conta Google conectada com sucesso',
        }),
      }
    } catch (error: any) {
      await prisma.integrationSettings.updateMany({
        where: { clinicId: state.clinicId },
        data: { gmailConnected: false },
      })

      const message = error instanceof AppError
        ? error.message
        : 'Falha ao concluir a autenticacao do Gmail'

      return {
        redirectUrl: appendRedirectParams(safeReturnUrl, {
          tab: 'integrations',
          gmail_oauth: 'error',
          gmail_message: message,
        }),
      }
    }
  }

  async getValidAccessToken(clinicId: string) {
    const settings = await prisma.integrationSettings.findUnique({
      where: { clinicId },
    })

    const credentials = getNormalizedGoogleCredentials(settings ?? {})
    const currentToken = settings?.gmailAccessToken
    const expiresAt = settings?.gmailTokenExpiresAt

    if (currentToken && (!expiresAt || expiresAt.getTime() > Date.now() + 60_000)) {
      return currentToken
    }

    if (!settings?.gmailRefreshToken) {
      throw new AppError('Gmail ainda nao autenticado. Faca a conexao OAuth primeiro.', 422)
    }

    const refreshed = await refreshAccessToken({
      clientId: credentials.clientId,
      clientSecret: credentials.clientSecret,
      refreshToken: settings.gmailRefreshToken,
    })

    await prisma.integrationSettings.update({
      where: { clinicId },
      data: {
        gmailClientId: credentials.clientId,
        gmailClientSecret: credentials.clientSecret,
        gmailAccessToken: refreshed.access_token,
        gmailTokenExpiresAt: refreshed.expires_in ? new Date(Date.now() + refreshed.expires_in * 1000) : null,
        gmailConnected: true,
      },
    })

    return refreshed.access_token
  }
}

File: backend/Consultorio.API/Services/GoogleOAuthService.cs
using System.IdentityModel.Tokens.Jwt;
using System.Net.Http.Headers;
using System.Security.Claims;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using Microsoft.EntityFrameworkCore;
using Microsoft.IdentityModel.Tokens;
using Consultorio.Domain.Models;
using Consultorio.Infra.Context;

namespace Consultorio.API.Services;

public class GoogleOAuthService
{
    private const string FallbackReturnPath = "/admin/configuracoes?tab=integrations";
    private static readonly string[] GmailScopes =
    [
        "https://www.googleapis.com/auth/gmail.readonly",
        "https://www.googleapis.com/auth/gmail.send",
        "https://www.googleapis.com/auth/gmail.modify",
    ];

    private readonly AppDbContext _db;
    private readonly IConfiguration _config;
    private readonly HttpClient _http;
    private readonly LegacyIntegrationBridge _legacyBridge;

    public GoogleOAuthService(
        AppDbContext db,
        IConfiguration config,
        HttpClient http,
        LegacyIntegrationBridge legacyBridge)
    {
        _db = db;
        _config = config;
        _http = http;
        _legacyBridge = legacyBridge;
    }

    private async Task<Clinic?> LoadClinicWithLegacyAsync(Guid clinicId)
    {
        var clinic = await _db.Clinics.FindAsync(clinicId);
        if (clinic == null)
            return null;

        if (await _legacyBridge.TryBackfillClinicAsync(clinic, LegacyIntegrationGroup.Gmail))
        {
            clinic.UpdatedAt = DateTime.UtcNow;
            await _db.SaveChangesAsync();
        }

        return clinic;
    }

    public async Task<(string AuthUrl, string RedirectUri)> CreateAuthorizationUrlAsync(
        HttpRequest request,
        Guid clinicId,
        Guid userId,
        string? returnUrl)
    {
        if (clinicId == Guid.Empty)
            throw new GoogleOAuthException("clinicId é obrigatório para iniciar a autenticação do Gmail", StatusCodes.Status400BadRequest);

        var hasAccess = await _db.SystemUsers.AnyAsync(su => su.ClinicId == clinicId && su.UserId == userId);
        if (!hasAccess)
            throw new GoogleOAuthException("Você não tem acesso a esta clínica", StatusCodes.Status403Forbidden);

        var clinic = await LoadClinicWithLegacyAsync(clinicId);
        if (clinic == null)
            throw new GoogleOAuthException("Clínica não encontrada", StatusCodes.Status404NotFound);

        if (string.IsNullOrWhiteSpace(clinic.GmailClientId) || string.IsNullOrWhiteSpace(clinic.GmailClientSecret))
            throw new GoogleOAuthException("Salve Client ID e Client Secret do Gmail antes de autenticar", StatusCodes.Status422UnprocessableEntity);

        var redirectUri = $"{ResolveRequestBaseUrl(request)}/api/auth/google/callback";
        var allowedOrigin = ResolveAllowedOrigin(request, returnUrl);
        var safeReturnUrl = BuildSafeReturnUrl(returnUrl, allowedOrigin);
        var stateToken = CreateStateToken(new GoogleOAuthStatePayload
        {
            ClinicId = clinicId,
            UserId = userId,
            ReturnUrl = safeReturnUrl,
            AllowedOrigin = allowedOrigin,
        });

        var query = new Dictionary<string, string>
        {
            ["client_id"] = clinic.GmailClientId,
            ["redirect_uri"] = redirectUri,
            ["response_type"] = "code",
            ["access_type"] = "offline",
            ["prompt"] = "consent",
            ["include_granted_scopes"] = "true",
            ["scope"] = string.Join(' ', GmailScopes),
            ["state"] = stateToken,
        };

        var authUrl = $"https://accounts.google.com/o/oauth2/v2/auth?{BuildQueryString(query)}";
        return (authUrl, redirectUri);
    }

    public async Task<string> HandleCallbackAsync(HttpRequest request)
    {
        var stateToken = request.Query["state"].FirstOrDefault();
        if (string.IsNullOrWhiteSpace(stateToken))
            throw new GoogleOAuthException("State OAuth ausente", StatusCodes.Status400BadRequest);

        var state = ReadStateToken(stateToken);
        var safeReturnUrl = BuildSafeReturnUrl(state.ReturnUrl, state.AllowedOrigin);

        var googleError = request.Query["error"].FirstOrDefault();
        var googleErrorDescription = request.Query["error_description"].FirstOrDefault();
        if (!string.IsNullOrWhiteSpace(googleError))
        {
            return AppendRedirectParams(safeReturnUrl, new Dictionary<string, string>
            {
                ["tab"] = "integrations",
                ["gmail_oauth"] = "error",
                ["gmail_message"] = googleErrorDescription ?? googleError,
            });
        }

        var code = request.Query["code"].FirstOrDefault();
        if (string.IsNullOrWhiteSpace(code))
        {
            return AppendRedirectParams(safeReturnUrl, new Dictionary<string, string>
            {
                ["tab"] = "integrations",
                ["gmail_oauth"] = "error",
                ["gmail_message"] = "Código OAuth do Google ausente",
            });
        }

        try
        {
            var hasAccess = await _db.SystemUsers.AnyAsync(su => su.ClinicId == state.ClinicId && su.UserId == state.UserId);
            if (!hasAccess)
                throw new GoogleOAuthException("Você não tem mais acesso a esta clínica", StatusCodes.Status403Forbidden);

            var clinic = await LoadClinicWithLegacyAsync(state.ClinicId);
            if (clinic == null)
                throw new GoogleOAuthException("Clínica não encontrada", StatusCodes.Status404NotFound);

            if (string.IsNullOrWhiteSpace(clinic.GmailClientId) || string.IsNullOrWhiteSpace(clinic.GmailClientSecret))
                throw new GoogleOAuthException("Credenciais OAuth do Gmail não configuradas para esta clínica", StatusCodes.Status422UnprocessableEntity);

            var redirectUri = $"{ResolveRequestBaseUrl(request)}/api/auth/google/callback";
            var tokens = await ExchangeCodeForTokensAsync(code, clinic.GmailClientId, clinic.GmailClientSecret, redirectUri);

            clinic.GmailAccessToken = tokens.AccessToken;
            clinic.GmailRefreshToken = !string.IsNullOrWhiteSpace(tokens.RefreshToken)
                ? tokens.RefreshToken
                : clinic.GmailRefreshToken;
            clinic.GmailTokenExpiresAt = tokens.ExpiresIn.HasValue
                ? DateTime.UtcNow.AddSeconds(tokens.ExpiresIn.Value)
                : null;
            try
            {
                clinic.GmailAddress = await GetProfileEmailAsync(tokens.AccessToken);
            }
            catch
            {
            }
            clinic.GmailConnected = true;
            clinic.UpdatedAt = DateTime.UtcNow;
            await _db.SaveChangesAsync();
            await _legacyBridge.TrySyncClinicAsync(clinic, LegacyIntegrationGroup.Gmail);

            return AppendRedirectParams(safeReturnUrl, new Dictionary<string, string>
            {
                ["tab"] = "integrations",
                ["gmail_oauth"] = "success",
                ["gmail_message"] = "Conta Google conectada com sucesso",
            });
        }
        catch (GoogleOAuthException ex)
        {
            await MarkClinicDisconnectedAsync(state.ClinicId);
            return AppendRedirectParams(safeReturnUrl, new Dictionary<string, string>
            {
                ["tab"] = "integrations",
                ["gmail_oauth"] = "error",
                ["gmail_message"] = ex.Message,
            });
        }
        catch
        {
            await MarkClinicDisconnectedAsync(state.ClinicId);
            return AppendRedirectParams(safeReturnUrl, new Dictionary<string, string>
            {
                ["tab"] = "integrations",
                ["gmail_oauth"] = "error",
                ["gmail_message"] = "Falha ao concluir a autenticação do Gmail",
            });
        }
    }

    public async Task<string> GetValidAccessTokenAsync(Guid clinicId)
    {
        var clinic = await LoadClinicWithLegacyAsync(clinicId);
        if (clinic == null)
            throw new GoogleOAuthException("Clínica não encontrada", StatusCodes.Status404NotFound);

        if (string.IsNullOrWhiteSpace(clinic.GmailClientId) || string.IsNullOrWhiteSpace(clinic.GmailClientSecret))
            throw new GoogleOAuthException("Credenciais do Gmail não configuradas", StatusCodes.Status422UnprocessableEntity);

        if (!string.IsNullOrWhiteSpace(clinic.GmailAccessToken) &&
            (!clinic.GmailTokenExpiresAt.HasValue || clinic.GmailTokenExpiresAt.Value > DateTime.UtcNow.AddMinutes(1)))
        {
            return clinic.GmailAccessToken;
        }

        if (string.IsNullOrWhiteSpace(clinic.GmailRefreshToken))
            throw new GoogleOAuthException("Gmail ainda não autenticado. Faça a conexão OAuth primeiro.", StatusCodes.Status422UnprocessableEntity);

        var refreshed = await RefreshAccessTokenAsync(
            clinic.GmailClientId,
            clinic.GmailClientSecret,
            clinic.GmailRefreshToken);

        clinic.GmailAccessToken = refreshed.AccessToken;
        clinic.GmailTokenExpiresAt = refreshed.ExpiresIn.HasValue
            ? DateTime.UtcNow.AddSeconds(refreshed.ExpiresIn.Value)
            : null;
        clinic.GmailConnected = true;
        clinic.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync();
        await _legacyBridge.TrySyncClinicAsync(clinic, LegacyIntegrationGroup.Gmail);

        return refreshed.AccessToken!;
    }

    public async Task<string> TestConnectionAsync(Guid clinicId)
    {
        var clinic = await LoadClinicWithLegacyAsync(clinicId);
        if (clinic == null)
            throw new GoogleOAuthException("Clínica não encontrada", StatusCodes.Status404NotFound);

        if (string.IsNullOrWhiteSpace(clinic.GmailClientId) || string.IsNullOrWhiteSpace(clinic.GmailClientSecret))
            throw new GoogleOAuthException("Salve o Client ID e Client Secret antes de testar", StatusCodes.Status422UnprocessableEntity);

        if (string.IsNullOrWhiteSpace(clinic.GmailAccessToken) && string.IsNullOrWhiteSpace(clinic.GmailRefreshToken))
            throw new GoogleOAuthException("Gmail salvo mas ainda não autenticado. Clique em \"Salvar e Autenticar\" para concluir o OAuth.", StatusCodes.Status422UnprocessableEntity);

        var accessToken = await GetValidAccessTokenAsync(clinicId);
        using var requestMessage = new HttpRequestMessage(HttpMethod.Get, "https://gmail.googleapis.com/gmail/v1/users/me/profile");
        requestMessage.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);

        var response = await _http.SendAsync(requestMessage);
        var body = await response.Content.ReadAsStringAsync();
        var profile = Deserialize<GoogleProfileResponse>(body);

        if (!response.IsSuccessStatusCode || string.IsNullOrWhiteSpace(profile?.EmailAddress))
        {
            clinic.GmailConnected = false;
            clinic.UpdatedAt = DateTime.UtcNow;
            await _db.SaveChangesAsync();
            await _legacyBridge.TrySyncClinicAsync(clinic, LegacyIntegrationGroup.Gmail);

            throw new GoogleOAuthException(
                ExtractGoogleError(body) ?? "Token inválido ou expirado",
                StatusCodes.Status400BadRequest);
        }

        clinic.GmailConnected = true;
        clinic.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync();
        await _legacyBridge.TrySyncClinicAsync(clinic, LegacyIntegrationGroup.Gmail);

        return profile.EmailAddress;
    }

    public async Task<string?> GetProfileEmailAsync(Guid clinicId)
    {
        var accessToken = await GetValidAccessTokenAsync(clinicId);
        return await GetProfileEmailAsync(accessToken);
    }

    private async Task<string?> GetProfileEmailAsync(string accessToken)
    {
        using var requestMessage = new HttpRequestMessage(HttpMethod.Get, "https://gmail.googleapis.com/gmail/v1/users/me/profile");
        requestMessage.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);

        var response = await _http.SendAsync(requestMessage);
        var body = await response.Content.ReadAsStringAsync();
        var profile = Deserialize<GoogleProfileResponse>(body);

        if (!response.IsSuccessStatusCode)
        {
            throw new GoogleOAuthException(
                ExtractGoogleError(body) ?? "Falha ao consultar o perfil do Gmail",
                StatusCodes.Status400BadRequest);
        }

        return profile?.EmailAddress;
    }

    private async Task<TokenResponse> ExchangeCodeForTokensAsync(string code, string clientId, string clientSecret, string redirectUri)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, "https://oauth2.googleapis.com/token")
        {
            Content = new FormUrlEncodedContent(new Dictionary<string, string>
            {
                ["code"] = code,
                ["client_id"] = clientId,
                ["client_secret"] = clientSecret,
                ["redirect_uri"] = redirectUri,
                ["grant_type"] = "authorization_code",
            })
        };

        var response = await _http.SendAsync(request);
        var body = await response.Content.ReadAsStringAsync();
        var token = Deserialize<TokenResponse>(body);

        if (!response.IsSuccessStatusCode || string.IsNullOrWhiteSpace(token?.AccessToken))
            throw new GoogleOAuthException(
                ExtractGoogleError(body) ?? "Falha ao trocar o código OAuth do Google",
                StatusCodes.Status400BadRequest);

        return token;
    }

    private async Task<TokenResponse> RefreshAccessTokenAsync(string clientId, string clientSecret, string refreshToken)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, "https://oauth2.googleapis.com/token")
        {
            Content = new FormUrlEncodedContent(new Dictionary<string, string>
            {
                ["client_id"] = clientId,
                ["client_secret"] = clientSecret,
                ["refresh_token"] = refreshToken,
                ["grant_type"] = "refresh_token",
            })
        };

        var response = await _http.SendAsync(request);
        var body = await response.Content.ReadAsStringAsync();
        var token = Deserialize<TokenResponse>(body);

        if (!response.IsSuccessStatusCode || string.IsNullOrWhiteSpace(token?.AccessToken))
            throw new GoogleOAuthException(
                ExtractGoogleError(body) ?? "Falha ao renovar o token do Gmail",
                StatusCodes.Status400BadRequest);

        return token;
    }

    private string CreateStateToken(GoogleOAuthStatePayload payload)
    {
        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_config["Jwt:Secret"]!));
        var credentials = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);

        var claims = new List<Claim>
        {
            new("type", "gmail_oauth"),
            new("clinicId", payload.ClinicId.ToString()),
            new("userId", payload.UserId.ToString()),
        };

        if (!string.IsNullOrWhiteSpace(payload.ReturnUrl))
            claims.Add(new Claim("returnUrl", payload.ReturnUrl));
        if (!string.IsNullOrWhiteSpace(payload.AllowedOrigin))
            claims.Add(new Claim("allowedOrigin", payload.AllowedOrigin));

        var token = new JwtSecurityToken(
            issuer: _config["Jwt:Issuer"],
            audience: _config["Jwt:Audience"],
            claims: claims,
            expires: DateTime.UtcNow.AddMinutes(15),
            signingCredentials: credentials);

        return new JwtSecurityTokenHandler().WriteToken(token);
    }

    private GoogleOAuthStatePayload ReadStateToken(string stateToken)
    {
        var parameters = new TokenValidationParameters
        {
            ValidateIssuer = true,
            ValidateAudience = true,
            ValidateLifetime = true,
            ValidateIssuerSigningKey = true,
            ValidIssuer = _config["Jwt:Issuer"],
            ValidAudience = _config["Jwt:Audience"],
            IssuerSigningKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_config["Jwt:Secret"]!)),
            ClockSkew = TimeSpan.FromMinutes(1),
        };

        var handler = new JwtSecurityTokenHandler();
        var principal = handler.ValidateToken(stateToken, parameters, out _);

        var type = principal.FindFirst("type")?.Value;
        if (type != "gmail_oauth")
            throw new GoogleOAuthException("State OAuth inválido", StatusCodes.Status400BadRequest);

        if (!Guid.TryParse(principal.FindFirst("clinicId")?.Value, out var clinicId))
            throw new GoogleOAuthException("State OAuth sem clínica válida", StatusCodes.Status400BadRequest);

        if (!Guid.TryParse(principal.FindFirst("userId")?.Value, out var userId))
            throw new GoogleOAuthException("State OAuth sem usuário válido", StatusCodes.Status400BadRequest);

        return new GoogleOAuthStatePayload
        {
            ClinicId = clinicId,
            UserId = userId,
            ReturnUrl = principal.FindFirst("returnUrl")?.Value,
            AllowedOrigin = principal.FindFirst("allowedOrigin")?.Value,
        };
    }

    private static string ResolveRequestBaseUrl(HttpRequest request)
    {
        var forwardedProto = request.Headers["X-Forwarded-Proto"].FirstOrDefault();
        var forwardedHost = request.Headers["X-Forwarded-Host"].FirstOrDefault();

        var protocol = (string.IsNullOrWhiteSpace(forwardedProto) ? request.Scheme : forwardedProto)
            .Split(',')[0]
            .Trim();
        var hostValue = string.IsNullOrWhiteSpace(forwardedHost) ? request.Host.Value : forwardedHost;
        var host = (hostValue ?? string.Empty)
            .Split(',')[0]
            .Trim();

        if (string.IsNullOrWhiteSpace(host))
            throw new GoogleOAuthException("Não foi possível determinar a URL pública da API", StatusCodes.Status500InternalServerError);

        return $"{protocol}://{host}";
    }

    private static string? ResolveAllowedOrigin(HttpRequest request, string? returnUrl)
    {
        var origin = request.Headers["Origin"].FirstOrDefault();
        if (Uri.TryCreate(origin, UriKind.Absolute, out var originUri))
            return originUri.GetLeftPart(UriPartial.Authority);

        var referer = request.Headers["Referer"].FirstOrDefault();
        if (Uri.TryCreate(referer, UriKind.Absolute, out var refererUri))
            return refererUri.GetLeftPart(UriPartial.Authority);

        if (Uri.TryCreate(returnUrl, UriKind.Absolute, out var returnUri))
            return returnUri.GetLeftPart(UriPartial.Authority);

        return null;
    }

    private static string BuildSafeReturnUrl(string? returnUrl, string? allowedOrigin)
    {
        if (string.IsNullOrWhiteSpace(returnUrl))
            return !string.IsNullOrWhiteSpace(allowedOrigin) ? $"{allowedOrigin}{FallbackReturnPath}" : FallbackReturnPath;

        if (returnUrl.StartsWith('/'))
            return !string.IsNullOrWhiteSpace(allowedOrigin) ? $"{allowedOrigin}{returnUrl}" : returnUrl;

        if (!Uri.TryCreate(returnUrl, UriKind.Absolute, out var parsed))
            return !string.IsNullOrWhiteSpace(allowedOrigin) ? $"{allowedOrigin}{FallbackReturnPath}" : FallbackReturnPath;

        if (!string.IsNullOrWhiteSpace(allowedOrigin) &&
            !string.Equals(parsed.GetLeftPart(UriPartial.Authority), allowedOrigin, StringComparison.OrdinalIgnoreCase))
        {
            return $"{allowedOrigin}{FallbackReturnPath}";
        }

        return parsed.ToString();
    }

    private static string AppendRedirectParams(string targetUrl, Dictionary<string, string> parameters)
    {
        var isAbsolute = Uri.TryCreate(targetUrl, UriKind.Absolute, out var absoluteUri);
        var uri = isAbsolute && absoluteUri != null
            ? absoluteUri
            : new Uri(new Uri("http://local"), targetUrl);

        var builder = new UriBuilder(uri);
        var queryValues = ParseQuery(builder.Query);
        foreach (var (key, value) in parameters)
            queryValues[key] = value;

        builder.Query = BuildQueryString(queryValues);
        if (isAbsolute)
            return builder.Uri.ToString();

        return $"{builder.Path}{(string.IsNullOrWhiteSpace(builder.Query) ? "" : $"?{builder.Query.TrimStart('?')}")}{builder.Fragment}";
    }

    private static Dictionary<string, string> ParseQuery(string? query)
    {
        var values = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        if (string.IsNullOrWhiteSpace(query))
            return values;

        foreach (var segment in query.TrimStart('?').Split('&', StringSplitOptions.RemoveEmptyEntries))
        {
            var parts = segment.Split('=', 2);
            var key = Uri.UnescapeDataString(parts[0]);
            var value = parts.Length > 1 ? Uri.UnescapeDataString(parts[1]) : "";
            values[key] = value;
        }

        return values;
    }

    private static string BuildQueryString(IReadOnlyDictionary<string, string> parameters) =>
        string.Join("&", parameters.Select(kvp =>
            $"{Uri.EscapeDataString(kvp.Key)}={Uri.EscapeDataString(kvp.Value ?? string.Empty)}"));

    private static T? Deserialize<T>(string json)
    {
        if (string.IsNullOrWhiteSpace(json))
            return default;

        try
        {
            return JsonSerializer.Deserialize<T>(json);
        }
        catch
        {
            return default;
        }
    }

    private static string? ExtractGoogleError(string? json)
    {
        if (string.IsNullOrWhiteSpace(json))
            return null;

        try
        {
            using var document = JsonDocument.Parse(json);
            var root = document.RootElement;

            if (root.TryGetProperty("error_description", out var errorDescription) &&
                errorDescription.ValueKind == JsonValueKind.String)
            {
                return errorDescription.GetString();
            }

            if (root.TryGetProperty("error", out var error))
            {
                if (error.ValueKind == JsonValueKind.String)
                    return error.GetString();

                if (error.ValueKind == JsonValueKind.Object &&
                    error.TryGetProperty("message", out var errorMessage) &&
                    errorMessage.ValueKind == JsonValueKind.String)
                {
                    return errorMessage.GetString();
                }
            }
        }
        catch
        {
        }

        return null;
    }

    private async Task MarkClinicDisconnectedAsync(Guid clinicId)
    {
        var clinic = await LoadClinicWithLegacyAsync(clinicId);
        if (clinic == null)
            return;

        clinic.GmailConnected = false;
        clinic.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync();
        await _legacyBridge.TrySyncClinicAsync(clinic, LegacyIntegrationGroup.Gmail);
    }

    private class GoogleOAuthStatePayload
    {
        public Guid ClinicId { get; set; }
        public Guid UserId { get; set; }
        public string? ReturnUrl { get; set; }
        public string? AllowedOrigin { get; set; }
    }

    private class TokenResponse
    {
        [JsonPropertyName("access_token")]
        public string? AccessToken { get; set; }

        [JsonPropertyName("refresh_token")]
        public string? RefreshToken { get; set; }

        [JsonPropertyName("expires_in")]
        public int? ExpiresIn { get; set; }
    }

    private class GoogleProfileResponse
    {
        [JsonPropertyName("emailAddress")]
        public string? EmailAddress { get; set; }
    }
}

public class GoogleOAuthException : Exception
{
    public int StatusCode { get; }

    public GoogleOAuthException(string message, int statusCode) : base(message)
    {
        StatusCode = statusCode;
    }
}
